File: 백준/Gold/1717. 집합의 표현/집합의 표현.cc
#include <iostream>
using namespace std;
int n, m;
int p[1000004];
// p[i] -> i의 부모, p[i]는 i의 부모가 누구에요???
int Find(int x){
    if(x == p[x])return x; // 부모 찾음
    else return p[x] = Find(p[x]); // 부모 못찾음
}

void Union(int x, int y){
    x = Find(x); // x의 부모
    y = Find(y); // y의 부모
    p[y] = p[x]; // y의 부모 <- x의 부모
}
int main(){
    ios_base::sync_with_stdio(false); cin.tie(NULL); cout.tie(NULL);

    cin >> n >> m;
    
    for(int i = 0; i <= n + 1; i++)p[i] = i;

    for(int i = 0; i < m; i++){
        int type, x, y;
		cin >> type >> x >> y;

        if (type == 0) {
			Union(x, y);
		}
		else {
			if (Find(x) == Find(y))cout << "YES\n";
			else cout << "NO\n";
		}
    }
    return 0;
}

File: 백준/Gold/1717. 집합의 표현/집합의 표현.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.nio.Buffer;
import java.util.StringTokenizer;

public class Main{

    static int n, m;
    static int[] p = new int[1000004];
    static int type, a, b;
    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());

        for(int i = 0; i <= n + 1; i++)p[i] = i;

        for(int i = 0; i < m; i++){
            st = new StringTokenizer(br.readLine());
            type = Integer.parseInt(st.nextToken());
            a = Integer.parseInt(st.nextToken());
            b = Integer.parseInt(st.nextToken());

            if(type == 0){
                Union(a, b);
            }else{
                if(Find(a) != Find(b)) System.out.println("NO");
                else System.out.println("YES");
            }
        }
    }
    static int Find(int x){

        if(x == p[x])return x;
        else return p[x] = Find(p[x]); // p[x]는 x의 부모
    }
    static void Union(int x, int y){

        x = Find(x);
        y = Find(y);
        p[y] = p[x]; // Y의 부모 <- X의 부모
    }
}